feat: Noports session logging

gkc opened the pull request:
**- What I did**
- Added `npevents` which listens for NoPorts session events (requested, approved / denied, daemonStarted, clientStarted, connected, ended). `npevents` takes a list of (policy service) atSigns with which it shares its event logging configuration
- Policy service new option `-l | --event-logging-atsign` and fetches the event logging config which that atSign has shared with it
- When daemon sends its heartbeat messages to Policy service, Policy sends its event logging config to the daemon. Daemon listens for messages from Policy and if it receives event logging config then it uses it from that point onwards.
- Session lifecycle
  - Client sends session request to daemon. Request now includes the relay atSign.
  - Daemon
    - emits `requested` session event
    - approves or denies request based on policy, and emits `approved` or `denied` session event accordingly
    - If session was approved, daemon sends notification to this session's relay atSign saying "use <this event logging config> for <this session>"
  - Relay
    - emits `connected` session event when the first pair of sockets has been joined together for this session
    - emits `ended` session event when the session ends; this event includes additional session info and statistics

**- How I did it**
See commits and diffs. I have added a bunch of comments on the diffs

**- How to verify it**
Tests pass

gkc committed:
feat: NoPorts events logging. Interim commit, comms are working end-to-end.

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

# Conflicts:
#	packages/dart/noports_core/lib/src/admin/impl.dart
#	packages/dart/noports_core/lib/src/common/at_ssh_key_util/local_ssh_key_util.dart
#	packages/dart/noports_core/lib/src/common/handle_server_events.dart
#	packages/dart/noports_core/lib/src/common/validation_utils.dart
#	packages/dart/noports_core/lib/src/npa/npa_impl.dart
#	packages/dart/noports_core/lib/src/npt/npt.dart
#	packages/dart/noports_core/lib/src/srv/srv.dart
#	packages/dart/noports_core/lib/src/srv/srv_impl.dart
#	packages/dart/noports_core/lib/src/srvd/isolates/port_pair_isolate.dart
#	packages/dart/noports_core/lib/src/srvd/isolates/relay_worker.dart
#	packages/dart/noports_core/lib/src/srvd/isolates/shared_single_port_isolate.dart
#	packages/dart/noports_core/lib/src/srvd/isolates/types.dart
#	packages/dart/noports_core/lib/src/srvd/relay_auth_verifiers.dart
#	packages/dart/noports_core/lib/src/srvd/srvd.dart
#	packages/dart/noports_core/lib/src/srvd/srvd_impl.dart
#	packages/dart/noports_core/lib/src/srvd/srvd_session_params.dart
#	packages/dart/noports_core/lib/src/sshnp/impl/sshnp_dart_pure_impl.dart
#	packages/dart/noports_core/lib/src/sshnp/impl/sshnp_openssh_local_impl.dart
#	packages/dart/noports_core/lib/src/sshnp/models/sshnp_arg.dart
#	packages/dart/noports_core/lib/src/sshnp/models/sshnp_params.dart
#	packages/dart/noports_core/lib/src/sshnp/models/sshnp_result.dart
#	packages/dart/noports_core/lib/src/sshnp/sshnp_core.dart
#	packages/dart/noports_core/lib/src/sshnp/util/sshnpd_channel/sshnpd_channel.dart
#	packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart
#	packages/dart/noports_core/test/srvd/notification_subscription_test.dart
#	packages/dart/noports_core/test/srvd/relay_authentication_legacy_test.dart
#	packages/dart/noports_core/test/srvd/srvdutil_test.dart
#	packages/dart/noports_core/test/sshnp/sshnp_core_test.dart
#	packages/dart/noports_core/test/sshnp/util/srvd_channel/srvd_channel_mocks.dart
#	packages/dart/noports_core/test/sshnp/util/srvd_channel/srvd_channel_test.dart

gkc committed:
chore: fix lint

gkc committed:
chore: dart format

gkc committed:
fix: oops missed a file

gkc committed:
fix: make relayAtsign optional in SshnpSessionRequest

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

# Conflicts:
#	packages/dart/sshnoports/lib/src/create_at_client_cli.dart

gkc committed:
build: update uuid dependency in npt_flutter

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

# Conflicts:
#	apps/admin/admin_api/lib/src/expose_apis.dart
#	packages/dart/noports_core/lib/src/sshnp/models/sshnp_arg.dart
#	packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart
#	packages/dart/npt_flutter/pubspec.lock
#	packages/dart/sshnoports/pubspec.lock

gkc committed:
feat: events logging: interim commit

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

gkc committed:
chore: restore docs/.gitbook/assets/SMB.png
chore: dart format
chore: remove unused import

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
[skip ci] chore: put files back the way they were

gkc committed:
refactor: rename Event* to AtEvent*
[skip ci]

gkc committed:
feat: all working after some fixes

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

gkc committed:
feat: all working after some fixes

gkc committed:
refactor: moved generic AtEvent stuff to at_client package

gkc committed:
feat: Add connector stats when sending the session ended event

gkc committed:
feat: overhauled the events mechanism to make integration with ELK / OpenSearch / etc easier
refactor: Brought the events stuff back from at_client, it's too immature.

[skip ci]

gkc committed:
feat: Updated the SessionEvents lifecycle stuff, fixed some bugs. Close to done now.

gkc committed:
build: add socket_connector dependency override to npt_flutter as well as the other packages

gkc committed:
build: add socket_connector dependency override to npt_flutter as well as the other packages

gkc committed:
Merge branch 'trunk' into noports-session-logging

gkc committed:
feat: Added compression/decompression to the AtEvent mixins
refactor: made AtEvent* mixins be `on AtClientBindings` for simplicity; renamed AtEventLoggingConfig to AtEventConfig
build(deps): added dependencies on the `shrink` and `at_base2e15` packages

gkc committed:
Merge remote-tracking branch 'origin/noports-session-logging' into noports-session-logging

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

gkc committed:
feat: `AtEventLogger` `logEvent` now writes to a stream which is processed asynchronously. So calling programs do not wait.
refactor: Renamed SessionLifecycle to SessionState; renamed experimental/event_listener.dart to npevents.dart; modified buildArchive accordingly
feat: added `stats` to the `newConnection` event
fix: ensured that all relevant SessionEvents are sent by sshnpd for sshnp sessions as well as npt
fix: improved `experimental/to.opensearch.sh`

gkc commented:
Still a work in progress but if you want to take this branch for a spin, do the following (MacOS / Linux)

- Build all the things
  - `cd /path/to/noports/repo/packages/dart/sshnoports`
  - `./buildArchive # The programs will be placed in build/sshnp, you will run them from there`
  - don't worry if the webapp build fails because you don't have node / npm - it's not relevant here
- Or
- Download the tarball from the latest multibuild for this branch

- Start local development instances of the opensearch and opensearch-dashboards services
  - `cd build/sshnp/experimental/opensearch`
  - `docker pull opensearchproject/opensearch:3`
  - `docker pull opensearchproject/opensearch-dashboards:3`
  - `docker compose up -d`
  - If you then run `docker compose ps` you should see output like this:
    ```
    gary@gkc2025 opensearch % docker compose ps
    NAME                    IMAGE                                            COMMAND                  SERVICE                 CREATED             STATUS              PORTS
    opensearch-dashboards   opensearchproject/opensearch-dashboards:latest   "./opensearch-dashbo…"   opensearch-dashboards   3 seconds ago       Up 2 seconds        0.0.0.0:5601->5601/tcp
    opensearch-node1        opensearchproject/opensearch:latest              "./opensearch-docker…"   opensearch-node1        3 seconds ago       Up 2 seconds        0.0.0.0:9200->9200/tcp, 9300/tcp, 0.0.0.0:9600->9600/tcp, 9650/tcp
    opensearch-node2        opensearchproject/opensearch:latest              "./opensearch-docker…"   opensearch-node2        3 seconds ago       Up 2 seconds        9200/tcp, 9300/tcp, 9600/tcp, 9650/tcp
    ```


- Start the new `npevents` service, piping its stdout to the `to.opensearch.sh` script
  - Note: It shares a logging config with the policy service's atSign. (You may use the same atSign for both)
  - `build/sshnp/experimental/npevents -a @eventsAtsign --logging-atsigns @policyAtsign --root-domain <whatever> --never-sync | build/sshnp/experimental/opensearch/to.opensearch.sh`
  - You should see output like this:
    ```
    Connecting ... Connected
    SHOUT|2025-10-12 14:49:37.622320| AtEventListenerService |Shared AtEventLogger config {atSign: @events, topic: 8K9eaFuB.events.logging.sshnp, ttln: 3600000} with [@policy] 
    State     Session
    =====     =======
    ```


- Start your policy service. It fetches the logging config from the events service's atSign. Note that it is fine to use the same atSign for both.
  - `build/sshnp/npp_atserver -a @policyAtsign -l @eventsAtsign --root-domain <whatever>`
  - You should see output like this:
    ```
    SHOUT|2025-10-12 14:53:31.144660|PolicyServiceWithAtClient|Loading groups via AtClient 
    SHOUT|2025-10-12 14:53:31.195201|PolicyServiceWithAtClient|Loading group from atKey: 1.groups.policy.sshnp@policy 
    SHOUT|2025-10-12 14:53:31.207327|PolicyServiceWithAtClient|Loaded 1.groups.policy.sshnp@policy - group name is (@client) 
    SHOUT|2025-10-12 14:53:31.207350|PolicyServiceWithAtClient|Load complete 
    SHOUT|2025-10-12 14:53:31.207357| npp |Daemon atSigns: {@service} 
    SHOUT|2025-10-12 14:53:31.341686| npp |Fetched AtEventLogger config {atSign: @events, topic: 8K9eaFuB.events.logging.sshnp, ttln: 3600000} from @events 
    SHOUT|2025-10-12 14:53:31.341712| npp |Starting AtRpc Server to listen for forced heartbeats... 
    SHOUT|2025-10-12 14:53:31.774548|PolicyServiceWithAtClient|Received device heartbeat from @service 
    SHOUT|2025-10-12 14:53:31.774596|NPAImpl|Sending config notification @service:config.mbp.devices.policy.sshnp@policy : {"eventLoggingConfig":{"atSign":"@events","topic":"8K9eaFuB.events.logging.sshnp","ttln":3600000}} 

    ```


- Start your relay(s) and daemon(s) as you normally do (but use the binaries from the `build/sshnp` directory)
  - the daemons should have the `-p @policyAtsign` option



- Run a NoPorts client session
- You should see output like this from the events to.opensearch.sh process:
  ```
  requested      e1cfba75-e110-48a4-afe0-711b30d55d0c
  approved       e1cfba75-e110-48a4-afe0-711b30d55d0c
  daemonStarted  e1cfba75-e110-48a4-afe0-711b30d55d0c
  clientStarted  e1cfba75-e110-48a4-afe0-711b30d55d0c
  connected      e1cfba75-e110-48a4-afe0-711b30d55d0c
  ended   e1cfba75-e110-48a4-afe0-711b30d55d0c

  ```


If the above has all worked then you can now view the events via the OpenSearch Dashboard
- Go to `http://localhost:5601`
- Click on the menu icon at the top left corner, and click `Discover`. You should see an "Index Patterns" screen. If not, paste this URL into your browser: `http://localhost:5601/app/management/opensearch-dashboards/indexPatterns`
- Click the "Create Index Pattern" button on the right
- Enter `noports` in the "Index pattern name" field and click "Next Step"
- Select the `timestamp` field as the primary time field and click "Create Index Pattern"
- Go back to the main "Discover" page. Ensure "noports" is selected in the dropdown field on the top left. You should see a screen like this:
<img width="1674" height="795" alt="image" src="https://github.com/user-attachments/assets/d31130ac-d1a1-4f30-b352-ed62d4b8dbfa" />

gkc committed:
feat: Added .env and docker-compose files to
packages/dart/sshnoports/bin/experimental/opensearch/

gkc commented:
I'll update the multibuild workflow tonight and run a multibuild against this branch

gkc committed:
feat: change `atsignA` to `clientAtsign` and `atsignB` to `daemonAtsign` in the event json
feat: include the device in the event json

gkc committed:
feat: small tweak to `to.opensearch.sh`

gkc committed:
docs: Add TODO to `to.opensearch.sh`

gkc committed:
refactor: move the experimental opensearch files into bundles/experimental/opensearch/ and update buildArchive accordingly
build: add npevents and opensearch bundle into multibuild.yaml

gkc commented:
> I'll update the multibuild workflow tonight and run a multibuild against this branch

Updated, and multibuild [completed](https://github.com/atsign-foundation/noports/actions/runs/18450317762) against commit ref [ad80fcd](https://github.com/atsign-foundation/noports/pull/2104/commits/ad80fcd0e904d207c98b44cba01e3c5408cdfeb7)

cconstab commented:
I got everything working but did not see the dashboard on localhost:5001

I just see

OpenSearch Dashboards server is not ready yet


I also noticed that the event scripted stopped working after re runing..

Perhaps we can sync and share screen tomorrow

Oh I also noticed that caching of policy on sshnpd means I think missing logs ?

cconstab commented:
After a but more playing I found that the docker setup needed more memory.. But still no logs in the dashboard... But I see the script sending them now (but interestingly no completed messages)


Speak tomorrow

cconstab commented:
Of course I had to have one last try..... I cleared the volumes and started open search (now I have more memory alolcated) and


<img width="1680" height="961" alt="Screenshot 2025-10-12 at 21 38 19" src="https://github.com/user-attachments/assets/10ada48e-da9b-4e61-b258-89baeaf3b2e0" />

cconstab commented:
The only thing I do not see is the session ending event with sshnp (I do with npt), but I get everything else :-)


sshnp
```
╰$ ./sshnp -f @alpha -t @bravo -r @romeo --root-domain vip.ve.atsign.zone -d test
2025-10-12 21:48:53.364662 : Sending daemon feature check request
2025-10-12 21:48:53.364718 : Resolving remote username for user session
2025-10-12 21:48:53.438891 : Resolving remote username for tunnel session
2025-10-12 21:48:53.439348 : Fetching host and port from srvd
2025-10-12 21:48:53.625140 : Received host and port from srvd
2025-10-12 21:48:53.625167 : Waiting for daemon feature check response
2025-10-12 21:48:53.699468 : Received daemon feature check response
2025-10-12 21:48:53.701755 : Required daemon features are supported
2025-10-12 21:48:53.864565 : Sending session request to the device daemon
2025-10-12 21:48:53.879327 : Waiting for response from the device daemon
2025-10-12 21:48:54.090603 : Received response from the device daemon
2025-10-12 21:48:54.093421 : Creating connection to socket rendezvous
2025-10-12 21:48:54.226708 : Starting user session
Last login: Sun Oct 12 21:47:17 2025 from ::1
╭─cconstab@tarial in ~
╰$
╭─cconstab@tarial in ~/Documents/GitHub/@foundation/sshnoports/packages/dart/sshnoports/build/sshnp on noports-session-logging ✘ (origin/noports-session-logging)
╰$
```

sshnpd
```
o
SHOUT|2025-10-12 21:47:17.332331| sshnpd |😎 Will handle ssh_request request from @alpha; acquired mutex 62e37013-6ccb-4545-9125-30d0d62c7ad0.session_mutexes.sshnp@bravo
SHOUT|2025-10-12 21:48:53.898317| sshnpd |😎 Will handle ssh_request request from @alpha; acquired mutex a7217ef5-ad9b-451b-ae68-862c94d6d17c.session_mutexes.sshnp@bravo
SHOUT|2025-10-12 21:50:32.695966| sshnpd |Config from policy: @bravo:config.test.devices.policy.sshnp@papa : {"eventLoggingConfig":{"atSign":"@echo","topic":"tgpgt86v.events.logging.sshnp","ttln":3600000}}
```

srvd
```
SHOUT|2025-10-12 21:48:53.577565| srvd main |Notification key: @romeo:test.request_ports.sshrvd@alpha
SHOUT|2025-10-12 21:48:53.601673| srvd main |Received ports (58985, 58986) in main isolate for session a7217ef5-ad9b-451b-ae68-862c94d6d17c
SHOUT|2025-10-12 21:48:53.601733| srvd main |😎 Will handle request from @alpha which can handle multiple acks (no mutex required)
SHOUT|2025-10-12 21:48:53.601744| srvd main |Started session a7217ef5-ad9b-451b-ae68-862c94d6d17c for @alpha to @bravo using ports (58985, 58986)
SHOUT|2025-10-12 21:48:53.601754| srvd main |Sending response data for requested session a7217ef5-ad9b-451b-ae68-862c94d6d17c : [127.0.0.1,58985,58986,2025-10-12T21:48:53.577659]
SHOUT|2025-10-12 21:49:23.603799| srvd / PortPairWorker / a7217ef5-ad9b-451b-ae68-862c94d6d17c |Finished session a7217ef5-ad9b-451b-ae68-862c94d6d17c for @alpha to @bravo using ports [58985, 58986]
```

but the scrip does not see the finished event

```
a7217ef5-ad9b-451b-ae68-862c94d6d17c requested
     Uploaded OK
a7217ef5-ad9b-451b-ae68-862c94d6d17c approved
     Uploaded OK
a7217ef5-ad9b-451b-ae68-862c94d6d17c daemonStarted
     Uploaded OK
a7217ef5-ad9b-451b-ae68-862c94d6d17c clientStarted
     Uploaded OK

```

gkc commented:
> but the script does not see the finished event

@cconstab found a bug here - daemon wasn't sending session logging config to the relay for requests from `sshnp`. Fixed, will push shortly.

gkc commented:
> > but the script does not see the finished event
> 
> @cconstab found a bug here - daemon wasn't sending session logging config to the relay for requests from `sshnp`. Fixed, will push shortly.

@cconstab Please note: I've changed the event names / states so they are now as follows:
```
State	Session
=====	=======
4dc50727-6eec-4fcc-ad52-6e478ef7cd51 requested
4dc50727-6eec-4fcc-ad52-6e478ef7cd51 approved
4dc50727-6eec-4fcc-ad52-6e478ef7cd51 daemonConnecting
4dc50727-6eec-4fcc-ad52-6e478ef7cd51 clientConnecting
4dc50727-6eec-4fcc-ad52-6e478ef7cd51 connected
4dc50727-6eec-4fcc-ad52-6e478ef7cd51 done
```

gkc committed:
fix: Ensure that session logging config is sent to relay for `sshnp` requests as well as for `npt` requests
refactor: Renamed session states to hopefully final form. (requested, denied, approved, daemonConnecting, clientConnecting, connected, done)
refactor: sshnpd: improved clarity of exception handling for startNpt, startDirectSsh, startReverseSsh; wrapped entirety of body of `clientRequestNotificationHandler` (the notification subscription handler) in a try-catch

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

gkc commented:
Have kicked off another [multibuild](https://github.com/atsign-foundation/noports/actions/runs/18466204428) for latest commits as of 13:45 UTC

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

# Conflicts:
#	packages/dart/npt_flutter/pubspec.yaml

gkc committed:
build(deps): Update socket_connector to 2.4.1

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

gkc committed:
chore: remove obsolete code from np_admin.dart

gkc committed:
refactor: Move some code around
[skip ci]

gkc committed:
build: add npevents in multibuild, dockerfiles

gkc committed:
test: e2e: main.sh now requires policyAtSign and eventsAtSign (not doing anything with them yet)
test: e2e: buildCurrentDartBinaries in common_functions_include.sh now also builds the npevents binary

gkc committed:
test: e2e: use staging instead of prod

gkc committed:
wat

gkc committed:
wat

gkc committed:
wat

gkc committed:
wat

gkc committed:
fix: handle atKeys secrets properly

gkc committed:
test: fix: set port as well as domain for atDirectory (C daemon requires host:port)

gkc committed:
test: fix: older c daemons seem buggy when using root-domain other than 'root.atsign.org'

gkc committed:
test: C daemon does not correctly parse port from host:port combo but instead ends up doing this `Connecting to root.atsign.wtf:64:64`

gkc committed:
Merge branch 'trunk' into noports-session-logging

gkc committed:
Merge branch 'trunk' into noports-session-logging

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

# Conflicts:
#	packages/dart/noports_core/lib/src/admin/impl.dart
#	packages/dart/noports_core/lib/src/sshnpd/sshnpd.dart
#	packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart
#	packages/dart/npt_flutter/pubspec.lock

gkc committed:
chore: resolve merge conflicts
[skip ci]

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

gkc commented on apps/admin/webapp/src/lib/PolicyUserGroups.svelte:
```diff
@@ -218,7 +218,7 @@
 
 <div class="row" style="color: {statusColor}; text-align: center">
-    <h2>{status}</h2>
+    <h2>{status}&nbsp;</h2>
```
drive-by fix to annoying UI glitch

gkc committed:
Compatibility improvements for the new event logging stuff.
- relay indicates to client if it supports event logging
- client only sends the relay atSign to the daemon if the relay supports event logging
- client sends a new flag to relay `sendJsonResponse` which asks the relay to send a json response rather than the legacy format. The client detects which style of response it receives from the relay, since older relays will continue to send the legacy format responses.
- some additional info logging when sending session requests to daemons
- Removed obsolete code related to relay session requests, improved null safety, added specific typing for relay response as well as relay request

gkc commented on .github/workflows/e2e_all.yaml:
```diff
@@ -35,19 +35,22 @@ jobs:
+          echo '${{ secrets.ATKEYS_NPE2E_RELAY }}' > ~/.atsign/keys/@npe2e_relay_key.atKeys
+          echo '${{ secrets.ATKEYS_NPE2E_RELAY_LATEST }}' > ~/.atsign/keys/@npe2e_relay_latest_key.atKeys
+          echo '${{ secrets.ATKEYS_NPE2E_EVENTS }}' > ~/.atsign/keys/@npe2e_events_key.atKeys
+          echo '${{ secrets.ATKEYS_NPE2E_SINGLETON }}' > ~/.atsign/keys/@npe2e_singleton_key.atKeys
```
We're using staging now. There's more to do here but that will be in a follow-up PR

gkc commented on apps/admin/admin_api/pubspec.lock:
```diff
@@ -745,14 +745,22 @@ packages:
       url: "https://pub.dev"
     source: hosted
     version: "3.0.0"
+  shrink:
```
For compressing what can be very large json messages in relay stats events. Have observed 93% size reduction for larger messages

xavierchanth replied:
I would prefer we use something more interoperable with other languages, shrink under the hood is zlib, but it seems it does some optimization magic which may potentially make it difficult to implement elsewhere.

gkc replied:
Thanks @XavierChanth good shout ... I've removed use of shrink and am just using straight zlib now - see [this commit](https://github.com/atsign-foundation/noports/pull/2104/commits/8d0bdf422e8d51df6aceee049a4864b4bacb999a). I also added some structure for forward flexibility - the event log notification payload is now a json-encoded envelope map which includes compressionScheme (default zlib), encodingScheme (base2e15), and the encodedCompressedPayload string

gkc commented on packages/dart/noports_core/lib/src/events/event_mixins.dart:
Two mixins to handle all of the boilerplate involved with event logging configs, logging of events and receipt of events.

Once we are happy with these mixins here, we can migrate them into the at_client package

gkc commented on packages/dart/noports_core/lib/src/events/noports_event_types.dart:
SessionState enum, plus a bunch of boilerplate-reducing statics to create the appropriate json messages

gkc commented on packages/dart/noports_core/lib/src/npa/npa.dart:
```diff
@@ -16,7 +16,7 @@ abstract class NPARequestHandler {
 /// - Checks if the clientAtSign is currently authorized to access
 ///   the sshnpd atSign and device
 /// - Responds accordingly
-abstract class NPA implements AtRpcCallbacks {
```
Created a separate class for the callbacks (see new class in npa_impl.dart, `PolicyInfoRpcRequestHandler`)

gkc commented on packages/dart/noports_core/lib/src/npa/npa_impl.dart:
```diff
@@ -91,42 +101,95 @@ class NPAImpl with AtClientBindings implements NPA {
-    AtRpc rpc = AtRpc(
+    _startPolicyInfoRpcServer();
+
+    subscribe(regex: r'.*\.devices\.policy\.sshnp', shouldDecrypt: true).listen(
```
Newer daemons will ping their policy service to get some additional config information. For now that's just `eventLoggingConfig` but this has been written to be extensible as and when we have additional config info to send

gkc commented on packages/dart/noports_core/lib/src/npa/npa_impl.dart:
```diff
@@ -91,42 +101,95 @@ class NPAImpl with AtClientBindings implements NPA {
+
+  late final AtRpc _policyInfoRpcServer;
+
+  void _startPolicyInfoRpcServer() {
```
Refactored this out into its own function

gkc commented on packages/dart/noports_core/lib/src/npa/npa_impl.dart:
```diff
@@ -91,42 +101,95 @@ class NPAImpl with AtClientBindings implements NPA {
   }
+}
+
+class PolicyInfoRpcRequestHandler
```
As mentioned above, moved the policy request handler into its own class rather than being bundled with NPA

gkc commented on packages/dart/noports_core/lib/src/npa/npa_impl.dart:
```diff
@@ -91,42 +101,95 @@ class NPAImpl with AtClientBindings implements NPA {
       'Received request from $fromAtSign: '
       '${jsonPrettyPrinter.convert(request.toJson())}',
     );
-    // We will send a 'log' notification to the loggingAtsign
```
Obsoleted

gkc commented on packages/dart/noports_core/lib/src/npa/npa_impl.dart:
```diff
@@ -151,9 +214,18 @@ class NPAImpl with AtClientBindings implements NPA {
         ).toJson(),
       );
     }
+    // We will send a 'log' notification to ourselves
```
This log notification is relevant only to the demo policy admin webapp and the noports desktop; it is not part of the new events logging stuff

gkc commented on packages/dart/noports_core/lib/src/npa/npa_params.dart:
```diff
@@ -55,6 +60,12 @@ class NPAParams {
+      'event-logging-atsign',
+      abbr: 'l',
+      help: 'atSign of a noports logging service.',
+    );
```
New command-line option

gkc commented on packages/dart/noports_core/lib/src/srvd/srvd_impl.dart:
```diff
@@ -182,19 +159,138 @@ class SrvdImpl implements Srvd {
     }
 
-    late SrvdSessionParams sessionParams;
+    logger.shout('Notification key: ${n.key}');
```
Additional code for handling the new notifications

gkc commented on packages/dart/noports_core/lib/src/srvd/srvd_impl.dart:
```diff
@@ -232,32 +328,35 @@ class SrvdImpl implements Srvd {
+            ..metadata = (Metadata()
+              ..immutable =
+                  true // only one srvd will succeed in doing this
+              ..ttl = 30000); // expire after 30 seconds to keep datastore clean
```
The above are format changes only

gkc commented on packages/dart/noports_core/lib/src/srvd/srvd_impl.dart:
```diff
@@ -386,6 +496,52 @@ class SrvdImpl implements Srvd {
     }
   }
 
+  Future<void> _handleSessionComplete(IIRequest msg) async {
```
Handlers here for messages which the isolates send regarding session events

gkc commented on packages/dart/noports_core/lib/src/srvd/srvd_impl.dart:
```diff
@@ -145,34 +150,6 @@ class SrvdImpl implements Srvd {
     initialized = true;
   }
 
-  Future<void> sendNack({
```
This function has been moved to later in this same file

gkc commented on packages/dart/noports_core/lib/src/srvd/srvd_session_params.dart:
```diff
@@ -1,157 +1,58 @@
-    return _sessionParamsForAncientClientRequest(notification);
-  }
-
-  /// Handles requests from ancient (v3) clients
```
We will no longer handle ancient clients

gkc commented on packages/dart/noports_core/lib/src/sshnp/util/srvd_channel/notification_request_message.dart:
Moved into relay_messages.dart (next file)

gkc commented on packages/dart/noports_core/lib/src/sshnp/util/srvd_channel/srvd_channel.dart:
```diff
@@ -210,22 +203,30 @@ abstract class SrvdChannel<T>
-      if (results.length >= 4) {
-        rvdNonce = results[3];
+
+      String notifVal = notification.value.toString();
```
If we get json response we construct RelayResponse from the json. If we get a legacy (non-json) response we construct a RelayResponse from the legacy format (comma separated list of values)

gkc commented on packages/dart/noports_core/lib/src/sshnp/util/srvd_channel/srvd_channel.dart:
```diff
@@ -235,67 +236,50 @@ abstract class SrvdChannel<T>
         ' rvdNonce: $rvdNonce',
       );
       logger.info('Daemon will connect to: $rvdHost:$daemonPort');
-      srvdAck = SrvdAck.acknowledged;
```
^^ was in the wrong place, moved up to before the `acked.complete()` above

gkc commented on packages/dart/noports_core/lib/src/sshnp/util/srvd_channel/srvd_channel.dart:
```diff
@@ -235,67 +236,50 @@ abstract class SrvdChannel<T>
+    );
 
-      rvdRequestValue = message.toString();
-    } else {
```
Ditching legacy request format

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -315,105 +325,132 @@ class SshnpdImpl with AtClientBindings, ApkamSigning implements Sshnpd {
-      if (notifPreProcessor != null) {
-        await notifPreProcessor!(notification);
+      try {
+        if (notifPreProcessor != null) {
```
put this check into its own try-catch block

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -315,105 +325,132 @@ class SshnpdImpl with AtClientBindings, ApkamSigning implements Sshnpd {
+    }
+  }
+
+  Future<void> _logEvent(Map<String, dynamic> event) async {
```
a wrapper so we don't have to have the same `if (elc != null)` check everywhere

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -627,6 +677,19 @@ class SshnpdImpl with AtClientBindings, ApkamSigning implements Sshnpd {
     }
 
+    await _logEvent(
+      SessionEvent.requested(
```
Session requested event

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -641,13 +704,25 @@ class SshnpdImpl with AtClientBindings, ApkamSigning implements Sshnpd {
     // Check if this *daemon* allows connections to the requested host / port
     if (!_permittedToOpen(permitOpen, req)) {
+      await _logEvent(
+        SessionEvent.denied(
```
session denied event

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -656,21 +731,68 @@ class SshnpdImpl with AtClientBindings, ApkamSigning implements Sshnpd {
     // Check if this *client* is allowed connections to the requested host / port
     if (!_permittedToOpen(auth.permitOpen, req)) {
+      await _logEvent(
+        SessionEvent.denied(sessionId: req.sessionId, authInfo: auth.toJson()),
```
denied

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -656,21 +731,68 @@ class SshnpdImpl with AtClientBindings, ApkamSigning implements Sshnpd {
         ),
-        value: 'Client is not permitted connections to $requested',
+        value:
+            'Connection to $requested denied based on POLICY --permit-open ${auth.permitOpen}',
```
better message

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -656,21 +731,68 @@ class SshnpdImpl with AtClientBindings, ApkamSigning implements Sshnpd {
     }
 
+    await _logEvent(
+      SessionEvent.approved(
```
session approved event

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -656,21 +731,68 @@ class SshnpdImpl with AtClientBindings, ApkamSigning implements Sshnpd {
+    );
+
     // Start our side of the tunnel
-    await startNpt(requestingAtsign: requestingAtsign, req: req);
```
To improve readability and maintainability, voved the try-catch handling out to here instead of within the startNpt function

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -656,21 +731,68 @@ class SshnpdImpl with AtClientBindings, ApkamSigning implements Sshnpd {
+      'relayAtsign ${req.relayAtsign}'
+      ' eventLoggingConfig $elc',
+    );
+    if (req.relayAtsign != null && elc != null) {
```
If the client sent us the relay atSign, and we got an event logging config from the policy service, then let's send the event logging config for this session to the relay

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -902,12 +1039,51 @@ class SshnpdImpl with AtClientBindings, ApkamSigning implements Sshnpd {
+    );
+
+    String which = req.direct ? 'startDirectSsh' : 'startReverseSsh';
+    try {
```
To improve readability and maintainability, voved the try-catch handling out to here instead of within the startDirectSsh / startReverseSsh functions

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -1596,28 +1764,82 @@ class SshnpdImpl with AtClientBindings, ApkamSigning implements Sshnpd {
+      '$messageType received from ${notification.from}:'
+      ' ${notification.value}',
+    );
+    switch (messageType) {
```
Some future proofing so that we can receive other types of messages from policy services in the future.

gkc commented on packages/dart/noports_core/lib/src/sshnpd/sshnpd_impl.dart:
```diff
@@ -1596,28 +1764,82 @@ class SshnpdImpl with AtClientBindings, ApkamSigning implements Sshnpd {
-          ..namespaceAware = true);
+    if (policyManagerAtsign == null) {
+      return;
+    }
```
better readability

gkc commented on packages/dart/noports_core/pubspec.yaml:
```diff
@@ -14,6 +14,8 @@ dependencies:
   at_cli_commons: ^3.0.1
   at_utils: ^3.0.19
+  at_base2e15: ^1.0.0
+  shrink: ^1.0.0
```
we use shrink to shrink the event payloads, and we encode the resulting bytes using base2e15

gkc commented on packages/dart/noports_core/pubspec.yaml:
```diff
@@ -22,7 +24,7 @@ dependencies:
   path: ^1.9.0
   posix: ^6.0.1
-  socket_connector: ^2.3.3
+  socket_connector: ^2.4.1
```
using new features in socket_connector

gkc commented on packages/dart/sshnoports/bin/npevents.dart:
Simplest possible program to listen to noports events and write them to stdout

gkc commented on packages/dart/sshnoports/bin/npevents.dart:
```diff
@@ -0,0 +1,128 @@
+      while (true) {
+        await for (final String json in svc.getJsonStream(elc)) {
+          stdout.writeln(json);
+        }
```
This is the actual code - get the events as a stream of json from the AtEventListenerService (cf the new files in noports_core/lib/src/events also in this PR)

gkc commented on packages/dart/sshnoports/bin/npp_atserver.dart:
```diff
@@ -82,11 +83,25 @@ void main(List<String> args) async {
+
+  AtEventConfig? elc;
+
+  if (p.eventLoggingAtsign != null) {
```
If we've got an event logging atSign then let's fetch the config that it has shared with us

gkc commented on tests/e2e_all/scripts/main.sh:
The changes below here are just some refactorings; the actual changes to the e2e tests (including running relays and policy services in their own containers) will follow in a subsequent PR

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

# Conflicts:
#	packages/dart/noports_core/lib/src/npa/npa_params.dart

gkc committed:
chore: fixed typo

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

gkc committed:
feat: removed dependency on `shrink` package, using `archive` package instead, supporting just zlib for now as a compression scheme
feat: added AtEventEnvelope to handle marshalling and unmarshalling the events being logged (maps) to the wire format (default being base2e15-encoded-zlib-compressed-json-encoding-of-the-map)
build(deps): Until at_client 3.9.3 is published, added dependency overrides for the `archive` package

gkc committed:
feat: better event messages for when a request is denied for POLICY reasons

gkc committed:
Merge remote-tracking branch 'origin/trunk' into noports-session-logging

xavierchanth commented on packages/dart/noports_core/lib/src/events/event_mixins.dart:
```diff
@@ -0,0 +1,317 @@
+    AtKey key = AtKey.fromString('$myAtsign:config.$namespace$atSign');
+    AtValue v = await atClient.get(key, getRequestOptions: _gro);
+    return AtEventConfig.fromJson(jsonDecode(v.value));
+  }
```
Having a static method on a mixin feels strange, especially as it causes there to be two ways of doing the same thing in this instance.

gkc replied:
Good catch thank you, yeah, was due to tech debt elsewhere which I'm now cleaning up.

gkc replied:
@XavierChanth changes done in [this commit](https://github.com/atsign-foundation/noports/pull/2104/commits/62665b099df72946b3c927c552b175c3f616c23b)

gkc committed:
refactor:
- removed obsolete `daemonAtSigns` instance variables / getters from various places
- moved some code from npp_atserver.dart and the NPA factory (fromCommandLineArgs) to its correct location in NPAImpl.run, with according changes to various functions' parameter lists etc
- and then removed the now-unnecessary `static staticGetEventLoggingConfig` from AtEventLogger mixin
chore: Changed some `shout` messages to `info`
